src/main/java/tech/stackable/hadoop/StackableTopologyProvider.java

```diff
@@ -175,144 +164,193 @@ private String getLabel(
-    Map<String, Map<String, String>> podLabels = getPodLabels(pods);
-    LOG.debug("Resolved podlabels for " + podLabels.keySet().toString());
+
+    Map<String, Map<String, String>> nodeLabels = getNodeLabels(datanodes);
+    LOG.debug("Resolved node labels map [{}]/[{}]", nodeLabels.keySet(), nodeLabels.values());
+
+    Map<String, Map<String, String>> podLabels = getPodLabels(datanodes);
+    LOG.debug("Resolved pod labels map [{}]/[{}]", podLabels.keySet(), podLabels.values());
+
+    names = resolveDataNodesFromCallingPods(names, podLabels, datanodes);
 
     // Iterate over all nodes to resolve and return the topology zones
-    for (String datanode : names) {
-      String builtLabel = getLabel(datanode, podLabels, nodeLabels);
+    for (String pod : names) {
+      String builtLabel = getLabel(pod, podLabels, nodeLabels);
       result.add(builtLabel);
 
       // Cache the value for potential use in a later request
-      this.topologyKeyCache.put(datanode, builtLabel);
+      this.topologyKeyCache.put(pod, builtLabel);
     }
+    LOG.info("Returning resolved labels [{}]", result);
     return result;
   }
 
   /**
-   * Given a list of Pods, return a HashMap that maps pod ips onto Pod labels. The returned Map may
-   * contain more entries than the list that is given to this function, as an entry will be
+   * The list of names may be datanodes (as is the case when the topology is initialised) or
+   * non-datanodes, when the data is being written by a client tool, spark executors etc. In this
+   * case we want to identify the datanodes that are running on the same node as this client. The
+   * names may also be pod IPs or pod names.
+   *
+   * @param names list of client pods to resolve to datanodes
+   * @param podLabels map of podIPs and labels
+   * @param dns list of datanode nameswhich will be used to match nodenames
+   * @return
+   */
+  private List<String> resolveDataNodesFromCallingPods(
+      List<String> names, Map<String, Map<String, String>> podLabels, List<Pod> dns) {
+    List<String> dataNodes = new LinkedList<>();
+    for (String name : names) {
+      // if we don't find a dataNode running on the same node as a non-dataNode pod, then
+      // we'll keep the original name to allow it to be resolved to NotFound in the calling routine.
+      String replacementDataNodeIp = name;
+      InetAddress address;
+      try {
+        // make sure we are looking up using the IP address
+        address = InetAddress.getByName(name);
+        String podIp = address.getHostAddress();
+        if (podLabels.containsKey(podIp)) {
+          replacementDataNodeIp = podIp;
+          LOG.info("added as found in the datanode map [{}]", podIp);
+        } else {
+          // we've received a call from a non-datanode pod
+          List<Pod> pods = client.pods().list().getItems();
+          for (Pod pod : pods) {
+            if (pod.getStatus().getPodIP().equals(podIp)) {
```

Comment:
Pods have a list of ips, might be better to use that and isin() instead of equals() to ensure finding all matches.

Reply:
Done: [3025fcc](https://github.com/stackabletech/hdfs-topology-provider/pull/5/commits/3025fcc1298c84c20a830d8fb1bb48e8bf00e564)